src/solidity/variable.js

```diff
@@ -0,0 +1,357 @@
+          if (dim[k] === '[]') {
+            ret.push('dynamic')
+          } else {
+            ret.push(parseInt(dim[k].replace('[', '').replace(']', '')))
+          }
+        }
+      }
+    }
+    return ret
+  },
+
+  /**
+  * retrieve enum declaration of the given @arg variable
+  *
+  * @param {Object} variable - variable declaration
+  * @param {Object} stateDeclarations  - all state declarations given by the AST (including struct and enum type declaration)
+  * @return [] containing all value declaration of the current enum type
+  */
+  getEnum: function (variable, stateDeclaration) {
+    for (var k in stateDeclaration) {
+      var dec = stateDeclaration[k]
+      if (dec.name === 'EnumDefinition' && variable.attributes.type.indexOf('enum ' + dec.attributes.name) === 0) {
+        return dec.children
+      }
+    }
+    return null
+  },
+
+  /**
+   * retrieve variable declared in the given @arg variable
+   *
+   * @param {Object} variable - variable declaration
+   * @param {Object} stateDeclarations  - all state declarations given by the AST (including struct and enum type declaration)
+   * @return [] containing all variables declaration of the current struct type
+   */
+  getStructMembers: function (variable, stateDeclaration) {
+    var ret = []
+    for (var k in stateDeclaration) {
+      var dec = stateDeclaration[k]
+      if (dec.name === 'StructDefinition' && variable.attributes.type.indexOf('struct ' + dec.attributes.name) === 0) {
+        var location = {
+          offset: 0,
+          slot: 0
+        }
+        for (var i in dec.children) {
+          var member = dec.children[i]
+          var type = this.getType(member, stateDeclaration)
+          var loc = this.walkStorage(type, location)
+          ret.push({
+            name: member.attributes.name,
+            type: type,
+            location: loc.currentLocation
+          })
+          location = loc.nextLocation
+        }
+        break
+      }
+    }
+    return {
+      members: ret,
+      slotsUsed: loc.nextLocation.slot
+    }
+  },
+
+  /**
+   * determine if the given @arg type need to start from a new slot
+   *
+   * @param {Object} type - current type (object returned by the getType function)
+   * @param {Object} location - location in storage of the type.
+   * @return true if the type needs a new storage
+   */
+  typeNeedNewSlot: function (type, location) {
+    return (type.size === 'dynamic' || location.offset + type.memSize > 64 || type.isArray || type.isStruct /* || type.isBytes*/)
+  },
+
+  /**
+   * determine what will be the location of the current @arg type and the start location of the next type in storage
+   *
+   * @param {Object} type - current type (object returned by the getType function)
+   * @param {Object} location - location in storage of the type.
+   * @return {Object} return the location of the current type and the location of the next type in storage
+   */
+  walkStorage: function (type, location) {
```

Comment:
Perhaps easier to understand if split into two functions:
`locationToFitType(type, location) -> location`
`locationToEnd(type, location) -> location`